### PR TITLE
Ported Nirvana to 0.2

### PR DESCRIPTION
**Pull request by jigold**

### CODE REVIEW

**Comment by danking** [comment]
@jigold what other PRs are in the stack?

File: python/hail/methods/qc.py [hunk 1]
```diff
@@ -169,3 +169,512 @@ def variant_qc(dataset, name='variant_qc'):
...
+    Here is an example `vep.properties` configuration file
+
+    .. code-block:: text
+
+        hail.vep.perl = /usr/bin/perl
+        hail.vep.path = /usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin
+        hail.vep.location = /path/to/vep/ensembl-tools-release-81/scripts/variant_effect_predictor/variant_effect_predictor.pl
+        hail.vep.cache_dir = /path/to/vep
+        hail.vep.lof.human_ancestor = /path/to/loftee_data/human_ancestor.fa.gz
+        hail.vep.lof.conservation_file = /path/to/loftee_data//phylocsf.sql
```
Comment (danking) [inline python/hail/methods/qc.py]: is the double forward slash intentional?

File: python/hail/methods/qc.py [hunk 2]
```diff
@@ -169,3 +169,512 @@ def variant_qc(dataset, name='variant_qc'):
...
+    - **hail.vep.plugin** -- VEP plugin, passed to VEP with the `--plugin`
+      option. Optional. Overrides `hail.vep.lof.human_ancestor` and
+      `hail.vep.lof.conservation_file`.
+    - **hail.vep.lof.human_ancestor** -- Location of the human ancestor file for
+      the LOFTEE plugin. Ignored if `hail.vep.plugin` is set. Required otherwise.
+    - **hail.vep.lof.conservation_file** -- Location of the conservation file
+      for the LOFTEE plugin. Ignored if `hail.vep.plugin` is set. Required
+      otherwise.
+
+    Here is an example `vep.properties` configuration file
```
Comment (danking) [inline python/hail/methods/qc.py]: Should this be double-asterisk to get fixed-with code? ([My source](http://www.sphinx-doc.org/en/stable/rest.html))

File: python/hail/methods/qc.py [hunk 3]
```diff
@@ -169,3 +169,512 @@ def variant_qc(dataset, name='variant_qc'):
...
+    - **hail.vep.fasta** -- Location of the FASTA file to use to look up the
+      reference sequence, passed to VEP with the `--fasta` option. Required.
+    - **hail.vep.assembly** -- Genome assembly version to use. Optional,
+      default: GRCh37
+    - **hail.vep.plugin** -- VEP plugin, passed to VEP with the `--plugin`
+      option. Optional. Overrides `hail.vep.lof.human_ancestor` and
+      `hail.vep.lof.conservation_file`.
+    - **hail.vep.lof.human_ancestor** -- Location of the human ancestor file for
+      the LOFTEE plugin. Ignored if `hail.vep.plugin` is set. Required otherwise.
+    - **hail.vep.lof.conservation_file** -- Location of the conservation file
```
Comment (danking) [inline python/hail/methods/qc.py]: My understanding is that all of these Hail field paths should be italics, which can, apparently, be achieved with single-asterisk or with single-backtick, though the meaning of single-backtick is unclear to me.

File: python/hail/methods/qc.py [hunk 4]
```diff
@@ -169,3 +169,512 @@ def variant_qc(dataset, name='variant_qc'):
...
+    following properties:
+
+    - **hail.vep.perl** -- Location of Perl. Optional, default: perl.
+    - **hail.vep.perl5lib** -- Value for the PERL5LIB environment variable when
+      invoking VEP. Optional, by default PERL5LIB is not set.
+    - **hail.vep.path** -- Value of the PATH environment variable when invoking
+      VEP.  Optional, by default PATH is not set.
+    - **hail.vep.location** -- Location of the VEP Perl script.  Required.
+    - **hail.vep.cache_dir** -- Location of the VEP cache dir, passed to VEP
+      with the `--dir` option. Required.
```
Comment (danking) [inline python/hail/methods/qc.py]: I think the references to CLI options should be double-backticks so as to render fixed-width.

File: python/hail/methods/qc.py [hunk 5]
```diff
@@ -169,3 +169,512 @@ def variant_qc(dataset, name='variant_qc'):
...
+
+    Notes
+    -----
+
+    **Configuration**
+
+    :py:meth:`~hail.VariantDataset.vep` needs a configuration file to tell it
+    how to run VEP. The format is a `.properties file
+    <https://en.wikipedia.org/wiki/.properties>`__. Roughly, each line defines a
+    property as a key-value pair of the form `key = value`. `vep` supports the
```
Comment (danking) [inline python/hail/methods/qc.py]: Should these both be double-backticks so as to render fixed-width? I feel more strongly about the name of binary executable being fixed-width.

File: python/hail/methods/qc.py [hunk 6]
```diff
@@ -169,3 +169,512 @@ def variant_qc(dataset, name='variant_qc'):
...
+    Add Nirvana annotations to the dataset:
+
+    >>> result = methods.nirvana(dataset, "data/nirvana.properties") # doctest: +SKIP
+
+    Notes
+    -----
+
+    ***Configuration***
+
+    :py:meth:`~hail.VariantDataset.nirvana` requires a configuration file. The
```
Comment (danking) [inline python/hail/methods/qc.py]: again, I think we're standardizing on `:meth:`

File: python/hail/methods/qc.py [hunk 7]
```diff
@@ -169,3 +169,512 @@ def variant_qc(dataset, name='variant_qc'):
...
+    format is a `.properties file
+    <https://en.wikipedia.org/wiki/.properties>`__, where each line defines a
+    property as a key-value pair of the form `key = value`. ``nirvana`` supports
```
Comment (danking) [inline python/hail/methods/qc.py]: I prefer an example from a plain text file (`key = value`) to use double-backticks.

File: python/hail/methods/qc.py [hunk 8]
```diff
@@ -169,3 +169,512 @@ def variant_qc(dataset, name='variant_qc'):
...
+    the following properties:
+
+    - **hail.nirvana.dotnet** -- Location of dotnet. Optional, default: dotnet.
+    - **hail.nirvana.path** -- Value of the PATH environment variable when
+      invoking Nirvana. Optional, by default PATH is not set.
+    - **hail.nirvana.location** -- Location of Nirvana.dll. Required.
+    - **hail.nirvana.reference** --Location of reference genome. Required.
+    - **hail.nirvana.cache** --Location of cache. Required.
+    - **hail.nirvana.supplementaryAnnotationDirectory** -- Location of
```
Comment (danking) [inline python/hail/methods/qc.py]: All of these should be italics

**Comment by jigold** [comment]
The VEP PR already got merged #2703.

**Comment by danking** [comment]
Ah, I think this PR needs a rebase then and I probably commented on a bunch of VEP changes

File: python/hail/methods/qc.py [hunk 9]
```diff
@@ -169,3 +169,512 @@ def variant_qc(dataset, name='variant_qc'):
...
+        --json
+        --everything
+        --allele_number
+        --no_stats
+        --cache --offline
+        --dir <hail.vep.cache_dir>
+        --fasta <hail.vep.fasta>
+        --minimal
+        --assembly <hail.vep.assembly>
+        --plugin LoF,human_ancestor_fa:$<hail.vep.lof.human_ancestor>,filter_position:0.05,min_intron_size:15,conservation_file:<hail.vep.lof.conservation_file>
```
[the remark this answers is not in the document]
Comment (konradjk) [reply python/hail/methods/qc.py]: Nope, no dollar sign there

File: python/hail/methods/qc.py (hunk 3 above)
Comment (jigold) [reply python/hail/methods/qc.py]: @danking these aren't hail field paths. They're keys in the properties file.

@tpoterba The format I have currently is what we decided upon for definition lists a year ago. We could change it to the definition list format (http://www.sphinx-doc.org/en/stable/rest.html#lists-and-quote-like-blocks).

Comment (danking) [reply python/hail/methods/qc.py]: Ah. I guess I'm ambivalent then. Using sphinx-style definition lists seems like a Good Thing (tm).

File: python/hail/methods/qc.py (hunk 8 above)
Comment (jigold) [reply python/hail/methods/qc.py]: I'm pushing back on this. I think it looks fine the way it is. We can revisit later if needed.

**Comment by danking** [comment]
@jigold needs rebase

**Pull request merged**